### PR TITLE
Adding PhaseCurve class for periodic system

### PR DESCRIPTION
**Pull request by labsaha**
This PR adds `PhaseCurve` class for calculating the phase of a periodic system at a given time. It also calculates the normalization factor at that phase of the system by ring interpolation using a given table of `phase` vs `normalization factor`.

### CODE REVIEW

File: gammapy/time/models.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+from ..utils.modeling import Parameter, ParameterList
+
+__all__ = [
+    'PhaseCurve',
+]
+
+log = logging.getLogger(__name__)
+
+
+class PhaseCurve(object):
```
Comment (cdeil) [inline gammapy/time/models.py]: @labsaha - all of the code in the class (including the docstring) should be indented.

File: gammapy/time/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import pytest
+from numpy.testing import assert_allclose
+from gammapy.time.models import PhaseCurve
+from gammapy.utils.testing import requires_data
+from gammapy.utils.scripts import make_path
+
+
+
+@pytest.fixture(scope='session')
+def phase():
```
Comment (cdeil) [inline gammapy/time/tests/test_models.py]: Please call this `phase_curve` since it is a `PhaseCurve` object.

File: gammapy/time/models.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    reference: The MJD value where phase is considered as 0.
+    phase0: phase at the reference MJD
+    f0:1st order derivative of the function phi with time
+    f1:2nd order derivative of the function phi with time
+    f2:3rd order derivative of the function phi with time
+    phase as function of time is computed using
+    Φ(t)=Φ0+f0(t−t0)+(1/2)f1(t−t0)^2+(1/6)f2(t−t0)^3
+
+
+    Returns
```
Comment (cdeil) [inline gammapy/time/models.py]: The class docstring should not have a returns section.
It has a Parameters section that documents the parameters of `__init__`.
`__init__` always returns an instance, doesn't contain a return statement and doesn't need a documentation of the return value.
Presumably here you're documenting the return value of one of the methods. That should be in the docstring of that method.

File: gammapy/time/models.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class PhaseCurve(object):
+    """PhaseCurve class.
+    
+    This is to calculate phase of a periodic system and to provide the 
+    normalization factor for the phase using circular interpolation. 
+    The required data for interpolation is given as an astropy table
+
+    # TODO: 1. Checking table values if the phases are in increasing order
```
Comment (cdeil) [inline gammapy/time/models.py]: Can you resolve the TODOs? Or explain what the problem is? Is it needed?

File: gammapy/time/models.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import logging
+import numpy as np
+from ..utils.modeling import Parameter, ParameterList
+
+__all__ = [
+    'PhaseCurve',
+]
+
+log = logging.getLogger(__name__)
```
Comment (cdeil) [inline gammapy/time/models.py]: There's no need to emit log messages from this file. Please remove this getLogger call and the import logging above.

File: gammapy/time/models.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (cdeil) [inline gammapy/time/models.py]: At the very top of each file in Gammapy we have the license line. Please add it.

File: gammapy/time/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+from gammapy.time.models import PhaseCurve
+from gammapy.utils.testing import requires_data
+from gammapy.utils.scripts import make_path
+
+
+
+@pytest.fixture(scope='session')
+def phase_curve():
+    filename = make_path('$GAMMAPY_EXTRA/test_datasets/phasecurve_LSI_DC.fits')
+    print(filename)
```
Comment (cdeil) [inline gammapy/time/tests/test_models.py]: remove print statement

File: gammapy/time/tests/test_models.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+from astropy.table import Table
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose
+from gammapy.time.models import PhaseCurve
```
Comment (cdeil) [inline gammapy/time/tests/test_models.py]: use relative imports for gammapy, like from all other files.

File: gammapy/time/models.py (hunk 4 above)
Comment (labsaha) [reply gammapy/time/models.py]: Actually, we don't need these TODOs. This was for some checks. Now I feel that it is not correct to put these checks. Hence, I remove TODOs from the docstring

**Comment by cdeil** [comment]
@labsaha - I've done some cleanup in e0946d3c .

I used `python setup.py build_docs` locally to check the formatting of the docstring. It now looks like this:

screen shot 2018-02-14 at 15 51 34

@labsaha - I would suggest to merge this PR now, and then to re-factor later once we have some "callers" for this code, i.e. start to work with these models and get some experience. OK?

**Comment by labsaha** [comment]
@cdeil The modification you did looks perfect! So we can merge this PR.

**Comment by cdeil** [comment]
@labsaha - I've done some more edits in fff5ca58, trying to make the evaluate calls more clear and modular / testable.

Overall I'm not convinced yet by what we have here, the main question is if / where we should use `astropy.time.Time` objects, or otherwise if we only use floats if they should be seconds (common for pulsars) or days (common for binaries), or if we use time-like quantities that have a unit attached.

This will need further thought and discussion by binary / pulsar people.

I'm merging this PR now, and then we will change / refactor as we get experience with time models.

**Comment by cdeil** [comment]
@labsaha - Congratulations on getting your first PR merged into Gammapy!
Anticipating this I already added you to http://gammapy.org/team.html yesterday.
:-)

**Comment by labsaha** [comment]
@cdeil Thanks a lot for all your help to reach this level!!!!!!! :-)

I have also this `time` issue in my mind about the time units. To make it clear about the model used in 1DC data, I already contacted Masha and Roberta to discuss this issue. Once we have a more clear idea on this, we can modify the code.

**Comment by labsaha** [comment]
@cdeil Regarding `time` units - after going through the XML files of for the binaries of 1 DC data challenge and some verification with the orbital phases of some binaries, I understood that to calculate the phase for a pulsar or binary the time should be only given as MJD. Of course, it is entirely due to the reference phase given in the XML file in units of  MJD.  However, for real data we should keep the possibility to use time as `astropy.time.Time`object such that user can decide how to provide the time to calculate the phase.

**Comment by cdeil** [comment]
@labsaha - Basically at this moment you're the only dev / user of this code, so it's up to you where to use Time objects. If you want to make changes to code or docs or tests that establish behaviour, you can any time send a new pull request, or you can also group changes to this class in your next PR where you add first code to compute phase curves, or at least use times and phase curves in some real application.

**Pull request merged**